src/net.cpp

```diff
@@ -1136,6 +1152,31 @@ void CConnman::CreateNodeFromAcceptedSocket(SOCKET hSocket,
         }
     }
 
+	// Currently, users can open multiple connections to a reachable node using the same IP address and a 
+	// different port and request blockchain data at each connection. This behavior has the following impacts:
+	//   (1) Overwhelm the reachable node and consume its bandwidth by requesting data and discarding it
+	//     (see video demonstrations here: https://www.dropbox.com/s/pvasua7lnepdgfq/Occupy_Connections.mp4?dl=0
```

Comment:
If these videos contain any relevant rationale for this PR, consider sharing that information as text instead (and then remove the links).

For security reasons I don't think many regular contributors to this project (which tend to be highly security/OPSEC conscious) would be willing to open a media file which is bit-by-bit controlled by an untrusted uploader. FWIW, these lists are pretty long/scary https://cve.mitre.org/cgi-bin/cvekey.cgi?keyword=vlc and https://cve.mitre.org/cgi-bin/cvekey.cgi?keyword=ffmpeg :)

The above is a general observation: I'm _not_ claiming that there is anything fishy with these specific MP4 files.

Reply:
@practicalswift Sure. Actually, we have a research paper in submission. The submission is double-blind and usually, reviewers do not have time to go through data. Hence, I just did a screen recording and uploaded it. 

Here is what happens in the two videos. (1) A lightweight scrip occupies all the incoming connections of our reachable node using the same IP address and multiple ports (there are 65K ports on a commodity computer. An adversary can potentially make 64K outgoing connections using a single machine). (2) The second video shows each of those incoming connections continuously requesting block headers from our reachable node and discarding them. As a result, the node's bandwidth is exhausted while the adversary does not have to store anything. The adversary just keeps requesting block headers through each connection and discards them. 


Also important to mention is that the idea came to mind since I do blockchain synchronization analysis (i.e, execute getpeerinfo to see who has the up-to-date chain). During that analysis, I found 33 incoming connections from the same IP address, and they remained connected for 21 hours. That's when I started looking into this.

Reply:
@practicalswift Additional caveat (alongside bandwidth exhaustion) is the risk of transaction deanonymization. I am currently conducting experiments on that. 

33 incoming connections from one IP address for 21 hours looked fishy. So I reached out to @am33r who deployed 1 connection per IP policy to stop that.

Reply:
So there could be a number of explanations why someone would want to do this.

One is that perhaps they simply don't know that multiple connections doesn't really add accuracy to transaction announcements. This synchronization of inbound poisson timers was only added a few releases ago, specifically in response to observing deanonymizers making multiple connections (from the same IP, or from multiple).

Another is to increase the strength of bandwidth or other resource exhaustion attacks. I believe that this indeed helps the attacker, but I'm not sure how much. Our protection against volumetric DoS attacks against individual nodes is low, and it should be almost as easily be possible to perform them with a single connection.

As for disallowing multiple connections from the same IP:
* When incoming transaction slots run out, we already prioritize peers from varied IP ranges, and so multiple connections from the same IP would already be preferred for disconnection. It works per IP range, because it's believed that in general it's cheap for attackers to get access to multiple IPs in the same range already.
* If we're actually observing attacks in the wild (especially effective DoS attacks), we should obviously do something against that; your solution is one, but it's fairly aggressive.
* A slightly less aggressive solution would be added to the inbound prioritization logic a feature that prefers disconnecting multiple connections from the same IP (even more than multiple connections from the same IP range).
* If we'd actually want to refuse like you propose here, it would need a bunch of safeguards (e.g. localhost would need to be excluded, because many local services may connect directly locally; also it should be able to bypass this policy using net permissions like NOBAN).